server/accounts.go

```diff
@@ -1355,43 +1355,47 @@ func (a *Account) subscribeInternal(subject string, cb msgHandler) (*subscriptio
 		return nil, fmt.Errorf("no internal account client")
 	}
 
-	sub, err := c.processSub([]byte(subject+" "+sid), false)
+	sub, err := c.processSub(createSub(c, subject, sid, cb), false)
 	if err != nil {
 		return nil, err
 	}
 
-	sub.icb = cb
 	return sub, nil
 }
 
 // This will add an account subscription that matches the "from" from a service import entry.
 func (a *Account) addServiceImportSub(si *serviceImport) error {
 	a.mu.Lock()
 	c := a.internalClient()
-	sid := strconv.FormatUint(a.isid+1, 10)
-	a.mu.Unlock()
-
 	// This will happen in parsing when the account has not been properly setup.
 	if c == nil {
+		a.mu.Unlock()
 		return nil
 	}
-
 	if si.sub != nil {
+		a.mu.Unlock()
 		return fmt.Errorf("duplicate call to create subscription for service import")
 	}
+	a.isid++
+	sid := strconv.FormatUint(a.isid, 10)
+	cb := func(sub *subscription, c *client, subject, reply string, msg []byte) {
+		c.processServiceImport(si, a, msg)
+	}
+	si.sub = createSub(c, si.from, sid, cb)
```

Comment:
Should createSub hang off of client versus always passing in as first arg?

Reply:
I did at first but thought it might be misleading. Wanted to be clear that this was just a helper instead of having &subscription{} everywhere. But I can certainly make it a `*client` receiver.

Reply:
Good point, could go either way.

Reply:
And to be clear: as of now it looks weird and really would be simpler to have `processSub(queue, sid string, cb msgHandler)` and create the subscription inside `processSub()`. And I could certainly do that. But in the mqtt branch I would have to add some more property to the subscription (namely `qos`). So again, could add `qos` as param to `processSub()`, or like I did in mqtt branch pass "subOptions" (or like here `*subscription` instead of introducing a new object type).
So to summarize, for main branch I could still solve the races but make it simple by just changing `processSub()` with some params and modify it only in the mqtt branch because of the need to have more properties in the `*subscription`. Wdyt?

Reply:
I like that idea, then as more progress is made on MQTT we can revisit.

Reply:
OK, pushed forced the changes. For the service import si.sub race, I would have had to keep the way it was, but realized that we were only using `si.sub` to get the `si.sub.sid` so I replaced `si.sub` with `si.sid` :-)